atom/browser/lib/power_observer_linux.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2017 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+#include "atom/browser/lib/power_observer_linux.h"
+
+#include <unistd.h>
+#include <uv.h>
+#include <iostream>
+
+#include "base/bind.h"
+#include "device/bluetooth/dbus/dbus_thread_manager_linux.h"
+
+const char kLogindServiceName[] = "org.freedesktop.login1";
+const char kLogindObjectPath[] = "/org/freedesktop/login1";
+const char kLogindManagerInterface[] = "org.freedesktop.login1.Manager";
+
+std::string get_executable_basename() {
+  char buf[4096];
+  size_t buf_size = 4096;
+  std::string rv("electron");
+  if (!uv_exepath(buf, &buf_size)) {
+    rv = strrchr(static_cast<const char*>(buf), '/') + 1;
+  }
+  return std::move(rv);
+}
+
+namespace atom {
+
+PowerObserverLinux::PowerObserverLinux()
+    : lock_owner_name_(get_executable_basename()), weak_ptr_factory_(this) {
+  auto dbus_thread_manager = bluez::DBusThreadManagerLinux::Get();
+  if (dbus_thread_manager) {
+    bus_ = dbus_thread_manager->GetSystemBus();
+    if (bus_) {
+      logind_ = bus_->GetObjectProxy(kLogindServiceName,
+                                     dbus::ObjectPath(kLogindObjectPath));
+      logind_->WaitForServiceToBeAvailable(
+          base::Bind(&PowerObserverLinux::OnLoginServiceAvailable,
+                     weak_ptr_factory_.GetWeakPtr()));
+    }
```

Comment:
We might want an else{} block here to warn on failure to get the system bus... though in practice this seems unlikely

Reply:
:+1: